spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/LazyCurrentTraceContext.java

```diff
@@ -0,0 +1,119 @@
+import org.apache.commons.logging.LogFactory;
+import org.springframework.beans.factory.BeanFactory;
+import org.springframework.context.ConfigurableApplicationContext;
+import reactor.util.annotation.Nullable;
+
+/**
+ * Spaghetti jar for various hacks around bean factory, mostly to avoid calling the
+ * expensive {@link BeanFactory#getBean(Class)} many times.
+ *
+ * <p>
+ * This should be used in code that must use {@link CurrentTraceContext}, but is wired
+ * prior to context refresh.
+ *
+ * <p>
+ * This defers calls to {@link ConfigurableApplicationContext#getBean(Class)} to
+ * accommodate not yet {@link ConfigurableApplicationContext#refresh() refreshed}
+ * contexts. It saves the result to avoid performance penalties.
+ */
+final class LazyCurrentTraceContext extends CurrentTraceContext {
+
+	// spring-jcl uses commons-logging, so do we.
+	private static final Log log = LogFactory.getLog(LazyCurrentTraceContext.class);
+
+	final BeanFactory beanFactory;
+
+	CurrentTraceContext currentTraceContext;
+
+	LazyCurrentTraceContext(BeanFactory beanFactory) {
+		this.beanFactory = beanFactory;
+	}
+
+	@Override
+	public TraceContext get() {
+		CurrentTraceContext latest = unwrap();
+		return latest != null ? latest.get() : null;
+	}
+
+	/**
+	 * Like {@link CurrentTraceContext#newScope(brave.propagation.TraceContext)}, except
+	 * it returns {@link Scope#NOOP} when the bean factory isn't ready.
+	 */
+	@Override
+	public Scope newScope(TraceContext currentSpan) {
+		CurrentTraceContext latest = unwrap();
+		return latest != null ? latest.maybeScope(currentSpan) : Scope.NOOP;
+	}
+
+	/**
+	 * Like {@link CurrentTraceContext#maybeScope(brave.propagation.TraceContext)}, except
+	 * it returns {@link Scope#NOOP} when the bean factory isn't ready.
+	 */
+	@Override
+	public Scope maybeScope(TraceContext currentSpan) {
+		CurrentTraceContext latest = unwrap();
+		return latest != null ? latest.maybeScope(currentSpan) : Scope.NOOP;
+	}
+
+	/**
+	 * Attempts to provision from the underlying bean factory, if not already provisioned.
+	 */
+	@Nullable
+	CurrentTraceContext unwrap() {
+		if (this.currentTraceContext != null) {
+			return this.currentTraceContext;
+		}
+
+		if (isActive()) {
+			if (log.isTraceEnabled()) {
+				log.trace("Spring context [" + beanFactory + "] already refreshed: "
+						+ "getting CurrentTraceContext.");
+			}
+
+			try {
+				this.currentTraceContext = beanFactory.getBean(CurrentTraceContext.class);
+			}
+			catch (Exception ex) {
+				if (log.isDebugEnabled()) {
+					log.debug("Spring context [" + beanFactory + "] error: "
+							+ "Will use NoOp CurrentTraceContext.", ex);
+				}
+			}
+		}
+		else if (log.isTraceEnabled()) {
+			log.trace("Spring context [" + beanFactory + "] not yet refreshed: "
+					+ "Will use NoOp CurrentTraceContext.");
+		}
+		return this.currentTraceContext;
+	}
+
+	boolean isActive() {
+		// TODO: Will Spring Boot ever return something besides ConfigurableApplicationContext?
```

Comment:
@wilkinsona @philwebb can you answer this one for me? I'd like to erase this question not just here, but everywhere.. with an assertion and blind cast, ideally, or a comment if it is actually expected.

Reply:
I don't think a `BeanFactory` will ever be a `ConfigurableApplicationContext`, at least not with Framework's various implementations of those two interfaces. Perhaps you could inject the context rather than the bean factory?

Much of Boot assumes that the context will be a `ConfigurableApplicationContext` (rooted in SpringApplication's requirement for it to be so). It is possible to use some pieces of Boot with a context that isn't a `ConfigurableApplicationContext`, but I would be surprised if many if any people do so. That said we do still have some `instanceof` checks in Boot's code. Some of those deal with parent contexts where you can be less sure of the type and I don't think you could reasonably assume it will be a `ConfigurableApplicationContext`.

Reply:
it looks like in sleuth a BeanDefinitionRegistryPostProcessor injects ConfigurableApplicationContext and passes it here as a BeanFactory https://github.com/spring-cloud/spring-cloud-sleuth/blob/76b7563feb58bbf31104a5c7dae080dcea469c06/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/TraceReactorAutoConfiguration.java#L141

However, this one injects a BeanFactory.. https://github.com/spring-cloud/spring-cloud-sleuth/blob/d08e106100c1e9957926517394097047a6884ea1/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceWebClientBeanPostProcessor.java#L149

I'll try to see if it will still work if all things just ask for the `ConfigurableApplicationContext`. While it will be a signature change, I'm not sure who would be using this utility besides us.. at worst, I guess we can do instanceof and overload it.

Reply:
since we injected both the narrowed type, ConfigurableApplicationContext, and also the wider type BeanFactory, it seems safe to assume we can narrow this to ConfigurableApplicationContext and still support what we did before. Since this is simpler, I'll do that in the next commit. Folks who want to deal with BF edge cases will want to ask about that. I think this was BF solely due to mocking anyway.

much appreciated background, @wilkinsona, I've copied the relevant bits into a comment.